add uefi variable store protocol

kraxel opened the pull request:
Initial bits of a UEFI variable store protocol implementation.

https://gitlab.com/kraxel/virt-firmware-rs/-/tree/main/varstore is a uefi
variable store implementation in rust.  Its far from being complete, but
should be good enough to get a linux guest boot up to the login prompt
(without secure boot).

The design idea is to re-use serialization protocol edk2 uses to talk to
SMM mode code.  The only change on the edk2 side is that the buffers are
passed to SVSM instead of SMM.  The same approach is used by the qemu
uefi variable store (new in qemu 10.0), which also receives request
buffers from the firmware.

So SVSM needs a simple protocol to send request buffers.  This is the
purpose of this PR.  Protocol documentation is in the `uefivars.md` file
created by the first patch.

TODO: Once merged use coconfs storage support to make uefi variables
persistent (will be a separate PR).

kraxel committed:
uefivars: add uefi mm protocol

Add protocol to receive mm communication requests from edk2.
Requests will be processed by virtfw_varstore crate which
implements an UEFI variable store.

The protocol is gated by the new 'uefivars' feature.  It is
enabled by default for the qemu builds.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
uefivars: add uefi mm protocol

Add protocol to receive mm communication requests from edk2.
Requests will be processed by virtfw_varstore crate which
implements an UEFI variable store.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

deeglaze commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,37 @@
+    let mut store = EfiVarStore::new();
+
+    // TODO: pass communication buffer addr+size from edk2.
+    // TODO: map communication buffer
```
Given the virtio-blk support in both OVMF and (soon) SVSM, is the plan to use virtqueues for logic reuse?

kraxel replied:
No.  The plan is to extend https://github.com/tianocore/edk2/tree/master/OvmfPkg/VirtMmCommunicationDxe so it can talk to both the qemu implementation and the svsm implementation.

kraxel commented:
> Number two is to setup the buffer once with a separate request (this is closer to the qemu device implementation):
> 
>     * SVSM_UEFI_MM_SETUP(paddr,size)
>     * SVSM_UEFI_MM_RESET
>     * SVSM_UEFI_MM_REQUEST

Picked this approach in the latest push.  Actual implementation is not there yet for the most part, but the intended API should be clear from the function declarations and comments.

kraxel commented:
The edk2 bits: https://github.com/tianocore/edk2/pull/11085

kraxel committed:
uefivars: write up uefi mm protocol spec draft

See added uefivars.md file for details.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
uefivars: write up uefi mm protocol spec draft

See added Documentation/docs/protocols/uefivars.md file for details.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

deeglaze commented on uefivars.md:
```diff
@@ -0,0 +1,94 @@
+ * MdeModulePkg/Include/Guid/SmmVariableCommon.h
+ * MdeModulePkg/Include/Guid/VarCheckPolicyMmi.h
+
+### linux refernces
```
```suggestion
### linux references
```

deeglaze commented on uefivars.md:
```diff
@@ -0,0 +1,94 @@
+communication between normal mode and management mode (MM for short).
+
+This SVSM protocol is a thin wrapper to allow edk2 firmware send those
+requests to the SVSM instead.  This allows the SVSM provide services
```
```suggestion
requests to the SVSM instead.  This allows the SVSM to provide services
```

deeglaze commented on uefivars.md:
````diff
@@ -0,0 +1,94 @@
+```
+preliminary, not yet officially assigned
+
+### query command
````
I think this should be in a section titled "calls" since each protocol has call numbers it specifies support for at different protocol versions.

deeglaze commented on uefivars.md:
````diff
@@ -0,0 +1,94 @@
+const SVSM_UEFI_MM_QUERY: u32 = 1;
+```
+
+Returns success in case the protocol is supported by the svsm.
````
I don't think we need to worry about adding a feature query call until there is a protocol version that needs the new feature and is optional in a dimension distinct from protocol version.

kraxel replied:
It's also meant as a way to probe whenever the protocol is present.

deeglaze replied:
The core protocol's query call is meant for that purpose.

stefano-garzarella replied:
@kraxel thanks for pinging me! I'll fix (or feel free to open a PR if you want)

deeglaze commented on uefivars.md:
````diff
@@ -0,0 +1,94 @@
+### setup command
+
+```
+const SVSM_UEFI_MM_SETUP: u32 = 2;
````
Setup is an overloaded term. Can we say this is REGISTER_BUFFER?

kraxel replied:
Yes.

kraxel commented on uefivars.md:
```diff
@@ -0,0 +1,94 @@
+
+Setup buffer for messages.
+
+Parameters:
+ * RCX: buffer address (guest physical).
+ * RDX: buffer size
```
Current edk2 side implementation uses page alignment (and 64k max size).

I'm fine with making page alignment a hard requirement.  Will also simplify the mapping code a bit.

Should we restrict the buffer size too?  Or allow the svsm restrict the size and throw some specific error in case the buffer is too big?

kraxel commented on uefivars.md:
````diff
@@ -0,0 +1,94 @@
+const SVSM_UEFI_MM_RESET: u32 = 3;
+```
+
+Reset protocol (communication buffer).
````
Well, this is meant to wipe any protocol state, right now this is more a UNREGISTER_BUFFER operation and probably wouldn't do much more in the future.  Guess we should rename this too.

kraxel commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    // map buffer
+    let start = paddr.page_align();
+    let offset = paddr.page_offset();
+    let end = (paddr + buffer.size).page_align_up();
+    let guard = PerCPUPageMappingGuard::create(start, end, 0)?;
+    let vaddr = guard.virt_addr() + offset;
+
+    // SAFETY: vaddr comes from a new mapped region.
+    let buffer = unsafe { from_raw_parts_mut(vaddr.as_mut_ptr::<u8>(), buffer.size) };
```
I'll have a look.

deeglaze replied:
There's additional drawback here from reading directly from the guest during processing when you might have validation to do ahead of time. It's much safer to copy in, copy out. This code is still missing additional safety checks for the whole region (you only check the initial address) and the requisite preconditions for creating a slice without UB. The copy in requirement is specifically to get a safety condition about the non-mutability of the slice from another source:

> The memory referenced by the returned slice must not be mutated for the duration of lifetime 'a, except inside an UnsafeCell.

So please use the library that was specifically written for handling guest memory reads and writes.

stefano-garzarella replied:
> Hmm, drawback is we'll go map the buffer twice per request then, first to read the request, then to write the response.

yeah, I was thinking the same when using that API in #603 , maybe we can try to create some abstraction. I'd suggest opening a PR with just that if you have something.

kraxel replied:
> yeah, I was thinking the same when using that API in #603 , maybe we can try to create some abstraction. I'd suggest opening a PR with just that if you have something.

It's actually three times.  First read fixed-size header, then read + write actual request (both variable size).  See most recent push which switches to guestmem functions (incremental patches for now, will clean up the branch once it passed testing).

tlendacky replied:
If the buffer will need to be mapped on each call anyway, does it makes sense to just have the single SVSM_UEFI_MM_REQUEST call which supplies the buffer address and size? It would be simpler and eliminate the need to register / unregister the buffer.

kraxel replied:
Well, the current protocol design gives the SVSM some flexibility how it wants handle the buffer mapping.  It can work with temporary (per cpu) mappings on each request, or it can work with a permanent (global) mapping.  Passing the buffer on each request takes the second option off the table.

If we declare we want stick to the temporary mappings forever then yes, we can go for a single REQUEST call.

kraxel committed:
uefivars: announce via query protocol

Update core protocol, query protocol request to
properly announce the uefi mm protocol.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
uefivars: announce via query protocol

Update core protocol, query protocol request to
properly announce the uefi mm protocol.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

tlendacky commented on uefivars.md:
```diff
@@ -0,0 +1,83 @@
+
+Parameters:
+ * RCX: buffer address (guest physical).  Must be page aligned.
+ * RDX: buffer size
```
Doh! right there in front of me.

tlendacky commented on uefivars.md:
```diff
@@ -0,0 +1,83 @@
+### request format
+
+The request serialization format is specified by edk2, the short
+overview below is purely informational.
```
Given that, do you want to stick to the register/unregister/request format or just have a single call with the buffer to be processed? I think the latter would be safer/better.

tlendacky replied:
Btw, what manifest information do you want to supply for this service for the attestation report?

deeglaze replied:
Since UEFI variables are going to be managed by SVSM, and it's up to the SVSM to provide protections of the persisted content, I think attesting to the persistence properties is prudent. Anything configurable and security-relevant needs to be measured. We're going to need to enable different behavior with flag flips as they become available and roll out according to production principles.

Now for what specifically should be attested, I think that we could use TCG non-volatile storage attributes TPMA_NV and TPMA_NV_EXP can be used as inspiration https://trustedcomputinggroup.org/wp-content/uploads/Trusted-Platform-Module-2.0-Library-Part-2-Version-184_pub.pdf 
I wouldn't recommend making the properties configurable at the per-variable level, so many bits are unnecessary.

deeglaze replied:
They are updateable by the UEFI specification with use of the appropriate keys. What you want to state is that the variable store does not permit SetVariable to non-volatile variables, then that should be attested. It might be plain from the behavior of the SVSM binary as measured while not yet configurable, so we could potentially wait to codify how these properties can be attested in a later manifest version. I'd much rather have some concretely reported, like nv_immutable for disallowing mutation of non-volatile UEFI variables.

kraxel replied:
Sketched a manifest, see the uefi_mm_get_manifest patch just pushed.

Writing non-volatile variables works fine, it is just that the changes are not actually persistent because we do not yet have encrypted storage support.

Writing non-volatile *authenticated* variables (including 'db' and 'dbx') is not supported yet due to lack of pkcs7 signature verification.  Also doesn't make much sense to support until we have persistence support for the efi variables.  Throws EFI_UNSUPPORTED errors right now.  Long-term we probably want a config switch to allow/deny that, and of course attest the state of the config switch.

tlendacky replied:
Also, this is the GUID I have assigned for the UEFI Management Mode protocol: a4453a59-9e1b-4787-a033-1986d6adbe55

kraxel replied:
> ```
> Byte Offset   Size (Bytes)   Meaning
> 0x000         4              Total size of manifest data
> 0x004         4              Flags
> 0x008         4              Offset of signature database (db)
> 0x00C         4              Size of signature database (db)
> 0x010         4              Offset of revoked signatures database (dbx) 
> 0x014         4              Size of revoked signatures database (dbx) 
> 0x018         4              Offset of Key Enrollment Key database (KEK) 
> 0x01C         4              Size of Key Enrollment Key database (KEK) 
> 0x020         Variable       Db / dbx / KEK 
> ```

Any specific reason for including the offsets and the total size?  These can be calculated from the size fields, and I'd prefer to not have redundant data in there.

Not sure whenever it makes sense to include the KEK.  For allowing / denying EFI binaries 'db' and 'dbx' are relevant.  If we want include KEK (because KEK signatures can update db+dbx) then we should include PK too (because PK signatures can update KEK).

deeglaze replied:
I don't think that including the variables in the manifest makes sense at all, given that they're measured into PCR7 when their contents are actually security-relevant.

tlendacky replied:
> I don't think that including the variables in the manifest makes sense at all, given that they're measured into PCR7 when their contents are actually security-relevant.

So just the flags field then?

kraxel replied:
> What is the format for db /dbx?

EFI signature database.

> Is there an identifier that would allow to distinguish which db is which?

No.

> We would need that documented somewhere in order to be sure this data can be parsed properly.

Yes, the order is important, and the size too.
(assuming we are adding these to the manifest, when relying on PCR7 instead this is moot ...).

kraxel committed:
uefivars: add uefi_mm_get_manifest

The manifest includes a version field and
a flags field with variable store properties.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
uefivars: add uefi_mm_get_manifest

The manifest includes a version field and
a flags field with variable store properties.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
[testing] uefivars: turn on secure boot

Flip switch.

RfC: What should be the default for the secureboot feature?

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
uefivars: wire up attestation manifest

Hook into attest_multiple_services() and
attest_single_service_handler().

Remove the now obsolete comment about
attest_single_service_handler() supporting vtpm only.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
uefivars: wire up attestation manifest

Hook into attest_multiple_services().

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

deeglaze commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,197 @@
+pub struct UefiMmManifestFlags {
+    // non-volatile uefi variables are written to persistent storage
+    pub persistent_nv_vars: bool,
+    // secure boot is enabled
+    pub secureboot_enabled: bool,
```
This is measured into PCR7. I wouldn't duplicate it here.

kraxel committed:
uefivars: add secure boot support

Enroll standard microsoft certificates ('db') and
revocation list ('dbx').  Also add a new 'secureboot'
feature to turn this on and off.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

kraxel committed:
uefivars: add secure boot support

Enroll standard microsoft certificates ('db') and
revocation list ('dbx').  Also add a new 'secureboot'
feature to turn this on and off.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

luigix25 commented on Documentation/docs/protocols/uefivars.md:
```diff
@@ -0,0 +1,70 @@
+
```
should this file be placed in in Documentation/docs/developer?

kraxel replied:
Does it make sense to have a new `Documentation/docs/protocol` directory?

luigix25 replied:
i think so!

kraxel replied:
ok.

luigix25 commented on uefivars.md:
````diff
@@ -0,0 +1,70 @@
+```
+pub const SVSM_UEFI_MM_PROTOCOL: u32 = 4;
+```
+preliminary, not yet officially assigned
````
is this still true?

Given https://github.com/tianocore/edk2/pull/11085, probably yes

kraxel replied:
No, it's in the spec and also merged already (commit e0e0e428d04ba247b331b9b4d2ccdac20c95993f).

I'll drop the line.

luigix25 commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,111 @@
+//!
+//! This SVSM protocol is a thin wrapper to allow edk2 firmware send
+//! those requests to the SVSM instead.  The actual request processing
+//! is implemented by the virtfw_varstore crate.  The later implements
```
/s/later/latter

kraxel replied:
fixed.

luigix25 commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,111 @@
+const SVSM_UEFI_MM_REQUEST: u32 = 1;
+
+// current edk2 uses 64k
+pub const UEFI_MM_BUFFER_LIMIT: usize = 256 * 1024;
```
does this need to be public?

kraxel replied:
no.

luigix25 commented on kernel/Cargo.toml:
```diff
@@ -86,6 +86,7 @@ noverify = []
 virtio-drivers = ["dep:virtio-drivers"]
 block = []
 uefivars = ["dep:virtfw-varstore"]
+secureboot = []
```
does secureboot need to depend on uefivars?

stefano-garzarella replied:
+1 on this.

Also I'm not sure if it makes sense to add a prefix like `uefivars-` ? Maybe no.

kraxel replied:
fixed

luigix25 commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -121,3 +126,41 @@ pub fn uefi_mm_protocol_init() -> Result<(), SvsmReqError> {
+struct UefiMmManifest {
+    pub version: u32,
+    pub flags: UefiMmManifestFlags,
+}
```
what about dropping dead_code and renaming fields to` _version` and `_flags`?

kraxel replied:
Well, the fields are used, so I wouldn't prefix them with `_`.  They get written, then the whole struct is converted to a byte vector.  They are never read as fields though, which why the rust compiler considers them unused.

luigix25 replied:
yeah I see. I don't have a strong opinion though.

luigix25 commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -121,3 +126,41 @@ pub fn uefi_mm_protocol_init() -> Result<(), SvsmReqError> {
+        .with_persistent_nv_vars(false)
+        .with_secureboot_enabled(sb)
+        .with_secureboot_db_update(false);
+    let manifest = UefiMmManifest { version: 0, flags };
```
why version 0?

kraxel replied:
Don't remember.  Guess I should better double-check the whole section against the latest svsm protocol spec by @tlendacky.

kraxel replied:
specs: https://docs.amd.com/v/u/en-US/58019.pdf

yes, version 0 is correct, and the other bits too.

luigix25 replied:
maybe worth a comment?

luigix25 commented on Cargo.toml:
```diff
@@ -93,6 +93,7 @@ serde = { version = "1.0.215", default-features = false }
 serde_json = { version = "1.0", default-features = false }
 sha2 = { version = "0.10.8", default-features = false }
 uuid = { version = "1.6.1", default-features = false }
+virtfw-libefi = { version = "0.6" }
```
latest version is 0.6.1, worth updating?

kraxel replied:
/me checks git log.

No.

luigix25 commented on Cargo.toml:
```diff
@@ -93,6 +93,7 @@ serde = { version = "1.0.215", default-features = false }
 sha2 = { version = "0.10.8", default-features = false }
 uuid = { version = "1.6.1", default-features = false }
+virtfw-libefi = { version = "0.6" }
 virtfw-varstore = { version = "0.6.1" }
```
ditto with 0.6.2

kraxel replied:
Also no.  No changes for the varstore core code.

stefano-garzarella commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,105 @@
+    let req = read_bytes_from_guest(paddr + boffset, bsize)?;
+    let rsp = core_request_dispatch(&mut store, &mmcore.guid, &req);
+    if rsp.len() > bsize {
+        // should not happen (add SvsmReqError::internal_error() ?)
```
yep, an assert in this case should be fine.

kraxel replied:
ok, switching to assert then.

stefano-garzarella commented on Documentation/docs/protocols/uefivars.md:
```diff
@@ -0,0 +1,70 @@
+   Communication Protocol" describes the EFI_MM_COMMUNICATE_HEADER
+   struct.
+
+### edk2 code references
+
+ * MdePkg/Include/Protocol/MmCommunication.h
+ * MdeModulePkg/Include/Guid/SmmVariableCommon.h
+ * MdeModulePkg/Include/Guid/VarCheckPolicyMmi.h
+
+### linux references
+
+The linux kernel has an efi variable driver implementation using this
+request format in `drivers/firmware/efi/stmm`.  It sends those
+requests to the edk2 MM code running in arm secure partition.
```
What about specifying from which versions they are available upstream in EDK2 and Linux?

kraxel replied:
I don't think this matters here.  Note that we have /two/ protocols here, the one is the serialization protocol for the requests (used in various places for years), and the other is the svsm protocol to pass those buffers between edk2 + svsm (which is very simple because it can piggyback on the existing serialization protocol).

The latter is not yet supported by edk2 even though I have a branch.  For the latter it'll make sense to document which edk2 version you'll need once it is merged.

stefano-garzarella commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,111 @@
+    let addr = params.rcx;
+    let size = params.rdx as usize;
+
+    let mut store = STORE.lock();
```
Do we really need to lock this spinlock from here?

What about this:
```diff
diff --git a/kernel/src/protocols/uefivars.rs b/kernel/src/protocols/uefivars.rs
index 00b54e51..b535c9e6 100644
--- a/kernel/src/protocols/uefivars.rs
+++ b/kernel/src/protocols/uefivars.rs
@@ -69,8 +69,6 @@ fn uefi_mm_request(params: &RequestParams) -> Result<(), SvsmReqError> {
     let addr = params.rcx;
     let size = params.rdx as usize;
 
-    let mut store = STORE.lock();
-
     // check buffer parameters
     check_buffer(addr, size)?;
     log::debug!("uefi mm buffer: 0x{addr:x} +0x{size:x}");
@@ -84,7 +82,7 @@ fn uefi_mm_request(params: &RequestParams) -> Result<(), SvsmReqError> {
     }
 
     let req = read_bytes_from_guest(paddr + boffset, bsize)?;
-    let rsp = core_request_dispatch(&mut store, &mmcore.guid, &req);
+    let rsp = core_request_dispatch(&mut STORE.lock(), &mmcore.guid, &req);
     if rsp.len() > bsize {
         // should not happen (add SvsmReqError::internal_error() ?)
         log::debug!("uefi mm: response buffer too big");
```

kraxel replied:
fixed.

stefano-garzarella commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -121,3 +126,41 @@ pub fn uefi_mm_protocol_init() -> Result<(), SvsmReqError> {
+
+#[allow(dead_code)]
+#[derive(IntoBytes, Immutable)]
+struct UefiMmManifest {
```
IIUC this will be parsed by edk2, so should we use `#[repr(C, packed)]` ?

kraxel replied:
It's not edk2, but yes, the struct layout is ABI.  I think for a struct with two `u32` it doesn't make an actual difference in struct layout, but it is good nevertheless to clarify our intention.

stefano-garzarella commented on kernel/src/protocols/attest.rs:
```diff
@@ -542,6 +546,10 @@ fn attest_multiple_services(params: &mut RequestParams) -> Result<(), SvsmReqErr
     services.push(SVSM_ATTEST_VTPM_GUID, vtpm_get_manifest()?);
+
+    #[cfg(all(feature = "uefivars", not(test)))]
+    services.push(SVSM_ATTEST_UEFI_MM_GUID, uefi_mm_get_manifest()?);
```
Should we add something also in `attest_single_service_handler()`?

kraxel replied:
Seems only the vtpm service is supported there right now.  If this makes sense for the uefi vars too (not sure it does) at minimum we'll need a guid assigned in the spec.

stefano-garzarella replied:
In the latest spec (https://docs.amd.com/v/u/en-US/58019.pdf - Revision: 1.01 - Issue Date: March 2026) I see:

> 10.2 Service Attestation Data
> 
> 10.2.1 Service Attestation GUID
> The UEFI Management Mode service attestation GUID is a4453a59-9e1b-4787-a033-1986d6adbe55.
> 
> 10.2.2 SVSM_ATTEST_SERVICES Manifest Data
> The manifest data used in an SVSM_ATTEST_SERVICES call is the same format as the manifest version 0 format of the SVSM_ATTEST_SINGLE_SERVICE call, see section 10.2.3.1.
> 
> 10.2.3 SVSM_ATTEST_SINGLE_SERVICE Manifest Data
> 
> 10.2.3.1 Manifest Version 0
> The manifest data used in an SVSM_ATTEST_SINGLE_SERVICE call for manifest version 0 is
> specified according to Table 45.

kraxel replied:
```
--- a/kernel/src/protocols/attest.rs
+++ b/kernel/src/protocols/attest.rs
@@ -583,6 +583,10 @@ fn attest_single_service_handler(params: &mut RequestParams) -> Result<(), SvsmR
     match attest_op.get_guid() {
         #[cfg(all(feature = "vtpm", not(test)))]
         SVSM_ATTEST_VTPM_GUID => attest_single_vtpm(params, &attest_op),
+        SVSM_ATTEST_UEFI_MM_GUID => {
+            attest_single_service(uefi_mm_get_manifest()?.as_slice(), params, &attest_op)
+        }
+
         _ => Err(SvsmReqError::unsupported_protocol()),
     }
 }
```
Compiles.  Not fully sure this actually makes sense.

stefano-garzarella replied:
> Compiles.

I guess the final code will need a guard on `uefivars` feature.

> Not fully sure this actually makes sense.

@kraxel What are your concerns?

kraxel replied:
If there is an actual use case for that.

Why does the single-service attestation exist in the first place?

kraxel replied:
Maybe I'm over-thinking this, included the change now.

stefano-garzarella replied:
IIUC was added to better support keylime use case where just the vTPM needed to be attested (see https://lore.kernel.org/linux-coco/6303283f-cf1c-8be6-9359-69d556a89554@amd.com/).

@tlendacky @jejb do you think makes sense to support SVSM_ATTEST_UEFI_MM_GUID also in the SVSM_ATTEST_SINGLE_SERVICE call or it is useless?

That said, I think we should also add the APIC Emulation at least in the multiple services version, but this is another issue.

stefano-garzarella commented on configs/qemu-target.json:
> [testing] uefivars: turn on secure boot
> Flip switch.
> 
> RfC: What should be the default for the secureboot feature?

Do we need to wait for the persistent storage or this can still be useful right now?

If it's already useful maybe we can discuss with the TSC what they prefer for now.

kraxel replied:
It really depends on the use case.  If you want boot standard distro kernels simply enrolling the standard microsoft certificates will work just fine.  If you want something more strict you'll have to configure svsm at compile time somehow because you can't (yet) boot the guest with a pre-configured variable store on coconfs.

kraxel replied:
Given that `cargo xbuild` has a cmd line switch to turn on features but not to turn off features it is probably most convinient to have it off by default ...

stefano-garzarella replied:
@kraxel thanks for the clarification. So that said, I agree that maybe it's better to turn it off by default and perhaps turn it on for tests (if we have tests running on that part)

kraxel replied:
dropped from config and turned on for test builds now.

kraxel committed:
uefivars: ensure test build coverage.

Add uefivars + secureboot to FEATURES_TEST.

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>

stefano-garzarella commented on configs/qemu-target.json:
```diff
@@ -14,7 +14,7 @@
     "kernel": {
         "svsm": {
-            "features": "vtpm",
+            "features": "vtpm,uefivars",
```
leftover (from https://github.com/coconut-svsm/svsm/pull/694#discussion_r3380638274) or intentional?

kraxel replied:
This turns on the variable store support by default but does not turn on secure boot.  Which I think makes sense.

stefano-garzarella replied:
@kraxel oh I see, makes sense to me too, maybe we can mention this in the commit.

stefano-garzarella commented on Documentation/docs/protocols/uefivars.md:
```diff
@@ -0,0 +1,69 @@
+
+## SVSM UEFI MM PROTOCOL
```
Should we mention that the `uefivars` feature must be enabled to support this? (somewhere in this file)

kraxel replied:
I'd tend to think the build instructions in `INSTALLATION.md` instructions would be the better place, especially as the edk2 side needs some care too.

Preparing the edk2 PR right now: https://github.com/tianocore/edk2/pull/12662

/me wants land this in the next (2026-08) edk2 release.

stefano-garzarella replied:
@kraxel yep, `INSTALLATION.md` seems a better place.

It would be nice to have some instructions also on how to test it for us, to check it's working during releases (I don't know, like some tools we can use in Linux to probe secureboot status, etc.)

kraxel replied:
Looking at the edk2 build build docs I guess it makes sense to rework them in a separate PR, they are a bit outdated.  svsm fork is not needed any more, all upstream in tianocore, also the grub/shim mess should be sorted meanwhile and 2026 linux distros should boot just fine without any hacks.  Will look into that.

stefano-garzarella replied:
@kraxel yep a follow up is fine by me. Thanks for that in advance!

stefano-garzarella commented on kernel/Cargo.toml:
```diff
@@ -82,6 +82,7 @@ virtio-drivers = ["dep:virtio-drivers"]
 block = []
 vsock = []
 uefivars = ["dep:virtfw-varstore"]
+secureboot = ["uefivars"]
```
Should we document this feature too in the new file we are adding the documentation to exaplain what this is supposed to do?

stefano-garzarella replied:
So this can go with the follow up as well.

stefano-garzarella commented on kernel/src/protocols/uefivars.rs:
```diff
@@ -0,0 +1,161 @@
+    _reserved: u32,
+}
+
+#[allow(dead_code)]
```
```suggestion
```

I just tried this change also with `make clippy CARGO_HACK=1` and I didn't see any error/warning. Do we really need it?

kraxel replied:
Hmm, interesting.  With `#[repr(C,packed)]` added it seems to not be needed any more.

stefano-garzarella replied:
@kraxel oh, cool, maybe with that it knows that it's an FFI and so fields can be unused in the rust code. Good to know!

stefano-garzarella reviewed (commented):
Some minor comments, mainly to improve a bit the documentation around the new added features.

The code LGTM!

stefano-garzarella commented on kernel/src/protocols/attest.rs:
I guess we can remove this comment now.

kraxel replied:
done

stefano-garzarella reviewed (dismissed):
LGTM, a minor minor comment left that we can fix only if you need to respin.

joergroedel reviewed (approved):
Really cool! Thanks @kraxel and @stefano-garzarella for bringing this over the finish line.
